server/lib/parties/party-service.ts

```diff
@@ -0,0 +1,217 @@
+        }),
+        () => {
+          // TODO(2Pac): Handle user quitting; need to keep a map of user -> invites?
+        },
+      )
+    })
+  }
+
+  removeInvite(partyId: string, target: PartyUser, leader?: PartyUser) {
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound)
+    }
+
+    if (leader && leader.id !== party.leader) {
+      throw new PartyServiceError(PartyServiceErrorCode.NotPartyLeader)
+    }
+
+    const updatedParty = party.deleteIn(['invites', target.id])
+    this.parties = this.parties.set(party.id as string, updatedParty)
+
+    const targetUser = this.getUser(target.name as string)
+    targetUser.unsubscribe(getInvitesPath(partyId))
+
+    this.publishToParty(partyId, {
+      type: 'decline',
+      target,
+    })
+
+    // TODO(2Pac): Do we want to remove the party record if there's only one member left in a party
+    // and no outstanding invites?
+  }
+
+  acceptInvite(partyId: string, user: PartyUser, clientId: string) {
+    const client = this.getClient(user.id as number, clientId)
+
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound)
+    }
+
+    if (party.members.size > MAX_PARTY_SIZE) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyFull)
+    }
+
+    const oldParty = this.getClientParty(client)
+    if (oldParty) {
+      // TODO(2Pac): Handle switching parties
+    }
+
+    const updatedParty = party.deleteIn(['invites', user.id]).setIn(['members', user.id], user)
+    this.parties = this.parties.set(party.id as string, updatedParty)
+    this.clientToPartyId = this.clientToPartyId.set(client, partyId)
+
+    this.publishToParty(partyId, {
+      type: 'join',
+      user,
+    })
+
+    const userSockets = this.getUser(user.name as string)
+    userSockets.unsubscribe(getInvitesPath(partyId))
+    this.subscribeToParty(client, updatedParty)
+  }
+
+  private subscribeToParty(client: ClientSocketsGroup, party: PartyRecord) {
+    client.subscribe(
+      getPartyPath(party.id as string),
+      () => ({
+        action: 'init',
+        party,
+      }),
+      client => this.handleClientQuit(client),
+    )
+  }
+
+  private handleClientQuit(client: ClientSocketsGroup) {
+    const party = this.getClientParty(client)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound)
+    }
+
+    this.clientToPartyId.delete(client)
+    this.parties = this.parties.deleteIn([party.id, 'members', client.userId])
+    this.publishToParty(party.id as string, {
+      type: 'leave',
+      user: client.userId,
+    })
+
+    // TODO(2Pac): Handle party leader leaving
+
+    // TODO(2Pac): Handle last person in a party leaving
+  }
+
+  private publishToParty(partyId: string, data: any) {
+    this.nydus.publish(getPartyPath(partyId), data)
+  }
+
+  private getUser(username: string): UserSocketsGroup {
```

Comment:
nit: I'd probably move to `getUserGroup` and `getClientGroup` (and use those names for variables as well) for this stuff, `user` is especially confusing since now it can mean:

 - a `User` object
 - a username
 - a user ID
 - a PartyUser
 - a UserSocketsGroup
 
If we can get that down to just one option (which in this context would be a PartyUser?) I think that'd be ideal :)

Reply:
What do you think about `getUserSockets` and `getClientSockets`? "Group" sounds ambiguous too imo, especially in a _party_ context :d.

But on the bright side, we now have TypeScript which makes stuff like this much less of a problem!

Reply:
That was my initial suggestion, actually, but it overlaps with what you're injecting the `UserSocketsManager` and `ClientSocketsManager` as :)

If you want to change that, that's fine.

And yeah, I agree it's nice that TS will prevent actual misuses, but it does make the code a bit harder to review at a glance, which is my main reason for wanting the rename